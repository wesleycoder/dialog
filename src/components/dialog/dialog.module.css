.dialog {
  position: fixed;
  top: 80vh;
  bottom: 0;
  max-width: 100vw;
  width: 100vw;
  height: 80vh;
  margin: 0;
  padding: 0;
  border: none;
  overflow: hidden;
  gap: 24px;
  transition: all .5s ease;
}

.dialog:not([open]) {
  pointer-events: none;
  opacity: 0;
}

.dialog::backdrop {
  background-color: rgba(0, 0, 0, 0.8);
  backdrop-filter: blur(2px);
  transition: backdrop-filter 5s ease;
}

.dialog:modal {
  display: flex;
  flex-direction: column;
}

.wrapper {
  display: flex;
  flex-direction: column;
  flex-grow: 1;
  padding: 32px;
  padding-bottom: 0;
  max-height: calc(100% - 32px);
}

.expanded {
  top: 20vh;
}

.dialog:has(*:hover) {
  top: 20vh;
}

.header {
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  font-size: 22px;
  line-height: 30px;
}

.close {
  margin: 0;
  padding: 0;
  border: none;
  background: none;
  cursor: pointer;
  padding-top: 8px;
  padding-left: 16px;
  padding-right: 16px;
  padding-bottom: 24px;
}

.closeIcon {
  width: 16px;
  height: 16px;
}

.title {
  margin: 0;
  font-weight: 500;
  font-size: 22px;
}

.content {
  display: flex;
  flex-grow: 1;
  overflow: auto;
}
