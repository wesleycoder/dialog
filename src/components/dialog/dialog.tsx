import {
  ForwardedRef,
  forwardRef,
  MouseEventHandler,
  PropsWithChildren,
  ReactNode,
  TouchEventHandler,
  useCallback,
  useImperativeHandle,
  useRef,
  useState,
} from 'react'
import { ReactComponent as IconClose } from '../../assets/icon-close.svg'
import styles from './dialog.module.css'

type DialogProps = PropsWithChildren<{ title: ReactNode }>

export type DialogHandle = {
  showModal(): void
} & Partial<HTMLDialogElement>

const DialogComponent = ({ title, children }: DialogProps, _ref: ForwardedRef<DialogHandle>) => {
  const [firstTouch, setFirstTouch] = useState<number>(0)
  const [lastTouch, setLastTouch] = useState<number>(0)
  const dialogRef = useRef<HTMLDialogElement>(null)

  useImperativeHandle(
    _ref,
    () => ({
      showModal() {
        if (dialogRef.current) {
          dialogRef.current.showModal()
        }
      },
      close() {
        if (dialogRef.current) {
          dialogRef.current.close()
        }
      },
    }),
    []
  )

  const closeDialog = useCallback(() => dialogRef.current?.close(), [])

  const handleClose: MouseEventHandler<HTMLDialogElement> = async e => {
    const isBackdrop = e.target instanceof Element && e.target.nodeName === 'DIALOG'

    if (isBackdrop) closeDialog()
  }

  const handleTouchMove: TouchEventHandler<HTMLDialogElement> = e => {
    if (!firstTouch) {
      setFirstTouch(e.targetTouches[0].clientY)
    }

    dialogRef.current?.style.setProperty('top', `${e.targetTouches[0].clientY}px`)
    dialogRef.current?.style.setProperty('transition', 'none')
    setLastTouch(e.targetTouches[0].clientY)
  }

  const handleTouchEnd: TouchEventHandler<HTMLDialogElement> = e => {
    dialogRef.current?.style.removeProperty('top')
    dialogRef.current?.style.removeProperty('transition')
    if (lastTouch <= firstTouch - 200) {
      dialogRef.current?.classList.add(styles.expanded)
    } else {
      dialogRef.current?.classList.remove(styles.expanded)
    }
    Promise.allSettled(e.currentTarget.getAnimations().map(animation => animation.finished)).then(() => {
      dialogRef.current?.style.removeProperty('top')
    })

    setFirstTouch(0)
    setLastTouch(0)
  }

  return (
    <dialog
      ref={dialogRef}
      className={styles.dialog}
      onClose={closeDialog}
      onClick={handleClose}
      onTouchMove={handleTouchMove}
      onTouchEnd={handleTouchEnd}
    >
      <div className={styles.wrapper}>
        <header className={styles.header}>
          <h2 className={styles.title}>{title}</h2>
          <button
            autoFocus
            className={styles.close}
            onClick={() => dialogRef.current?.close()}
          >
            <IconClose className={styles.closeIcon} />
          </button>
        </header>
        <div className={styles.content}>{children}</div>
      </div>
    </dialog>
  )
}

export const Dialog = forwardRef(DialogComponent)
