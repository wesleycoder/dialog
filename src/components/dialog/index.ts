export * from './dialog'
export { Dialog as default } from './dialog'
