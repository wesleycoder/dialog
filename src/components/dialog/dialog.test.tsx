import { describe, expect, it, vi } from 'vitest'
import { Dialog, DialogProps } from './dialog'
import { render } from '/config/test/render'

describe('Dialog', () => {
  it('should render hidden if isOpen is false', async () => {
    const props: DialogProps = {
      title: 'Lemon ipsum',
      isOpen: false,
      closeOnOverlayClick: false,
      onClose: vi.fn(),
    }

    const { container, findByRole } = render(<Dialog {...props} />)

    const dialog = (await findByRole('dialog', { hidden: true })) as HTMLDialogElement
    expect(dialog.open).toBe(props.isOpen)

    expect(container).toMatchSnapshot()
  })

  it('should render visible if isOpen is true', async () => {
    const props: DialogProps = {
      title: 'Lemon ipsum',
      isOpen: true,
      closeOnOverlayClick: false,
      onClose: vi.fn(),
    }

    const { container, findByRole, getByRole } = render(<Dialog {...props} />)

    const dialog = (await findByRole('dialog')) as HTMLDialogElement
    expect(dialog.open).toBe(props.isOpen)

    const header = getByRole('heading', { name: props.title })
    expect(header.textContent).toBe(props.title)

    expect(container).toMatchSnapshot()
  })

  it('should render with children', async () => {
    const props: DialogProps = {
      title: 'Lemon ipsum',
      isOpen: true,
      closeOnOverlayClick: false,
      onClose: vi.fn(),
    }

    const { container, getByRole, findByRole } = render(
      <Dialog {...props}>
        <p>Lemon ipsum dolor sit amet</p>
      </Dialog>
    )

    const dialog = (await findByRole('dialog')) as HTMLDialogElement
    expect(dialog.open).toBe(props.isOpen)

    const header = getByRole('heading', { name: props.title })
    expect(header.textContent).toBe(props.title)

    expect(container).toMatchSnapshot()
  })

  /**
   * FIXME: jsdom doesn't fully support HTMLDialogElement and close event still tricky to mock
   * watch for this issue to be resolved:
   * https://github.com/jsdom/jsdom/issues/3294
   */
  it.skip('should trigger onClose when clicking on close button', async () => {
    const props: DialogProps = {
      title: 'Lemon ipsum',
      isOpen: true,
      closeOnOverlayClick: false,
      onClose: vi.fn(),
    }

    const { getByRole, findByRole } = render(<Dialog {...props} />)

    const dialog = (await findByRole('dialog')) as HTMLDialogElement
    expect(dialog.open).toBe(props.isOpen)

    const closeButton = getByRole('button', { name: 'close' })

    closeButton.click()

    expect(props.onClose).toHaveBeenCalled()
  })
})
