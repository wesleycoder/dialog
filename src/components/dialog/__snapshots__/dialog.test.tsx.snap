// Vitest Snapshot v1

exports[`Dialog > should render hidden if isOpen is false 1`] = `
<div>
  <dialog
    class="_dialog_40d523"
  >
    <div
      class="_wrapper_40d523"
    >
      <header
        class="_header_40d523"
      >
        <h2
          class="_title_40d523"
        >
          Lemon ipsum
        </h2>
        <button
          aria-label="close"
          class="_close_40d523"
        >
          <svg
            class="_closeIcon_40d523"
            fill="none"
            height="16"
            role="img"
            viewBox="0 0 16 16"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M15 1L8 8L15 15"
              fill="none"
              stroke="currentColor"
              stroke-width="1.2"
            />
            <path
              d="M1 1L8 8L0.999999 15"
              fill="none"
              stroke="currentColor"
              stroke-width="1.2"
            />
          </svg>
        </button>
      </header>
      <div
        class="_content_40d523"
      />
    </div>
  </dialog>
</div>
`;

exports[`Dialog > should render visible if isOpen is true 1`] = `
<div>
  <dialog
    class="_dialog_40d523"
    open=""
  >
    <div
      class="_wrapper_40d523"
    >
      <header
        class="_header_40d523"
      >
        <h2
          class="_title_40d523"
        >
          Lemon ipsum
        </h2>
        <button
          aria-label="close"
          class="_close_40d523"
        >
          <svg
            class="_closeIcon_40d523"
            fill="none"
            height="16"
            role="img"
            viewBox="0 0 16 16"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M15 1L8 8L15 15"
              fill="none"
              stroke="currentColor"
              stroke-width="1.2"
            />
            <path
              d="M1 1L8 8L0.999999 15"
              fill="none"
              stroke="currentColor"
              stroke-width="1.2"
            />
          </svg>
        </button>
      </header>
      <div
        class="_content_40d523"
      />
    </div>
  </dialog>
</div>
`;

exports[`Dialog > should render with children 1`] = `
<div>
  <dialog
    class="_dialog_40d523"
    open=""
  >
    <div
      class="_wrapper_40d523"
    >
      <header
        class="_header_40d523"
      >
        <h2
          class="_title_40d523"
        >
          Lemon ipsum
        </h2>
        <button
          aria-label="close"
          class="_close_40d523"
        >
          <svg
            class="_closeIcon_40d523"
            fill="none"
            height="16"
            role="img"
            viewBox="0 0 16 16"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M15 1L8 8L15 15"
              fill="none"
              stroke="currentColor"
              stroke-width="1.2"
            />
            <path
              d="M1 1L8 8L0.999999 15"
              fill="none"
              stroke="currentColor"
              stroke-width="1.2"
            />
          </svg>
        </button>
      </header>
      <div
        class="_content_40d523"
      >
        <p>
          Lemon ipsum dolor sit amet
        </p>
      </div>
    </div>
  </dialog>
</div>
`;
